Zend/zend_string.c

```diff
@@ -72,175 +57,207 @@ ZEND_KNOWN_STRINGS(_ZEND_STR_DSC)
-			Bucket *old_buckets = CG(interned_strings).arData;
-
-			CG(interned_strings).nTableSize += CG(interned_strings).nTableSize;
-			CG(interned_strings).nTableMask = -CG(interned_strings).nTableSize;
-			new_data = malloc(HT_SIZE(&CG(interned_strings)));
-
-			if (new_data) {
-				HT_SET_DATA_ADDR(&CG(interned_strings), new_data);
-				memcpy(CG(interned_strings).arData, old_buckets, sizeof(Bucket) * CG(interned_strings).nNumUsed);
-				free(old_data);
-				zend_hash_rehash(&CG(interned_strings));
-			} else {
-				CG(interned_strings).nTableSize = CG(interned_strings).nTableSize >> 1;
-				CG(interned_strings).nTableMask = -CG(interned_strings).nTableSize;
-			}
+	return NULL;
+}
+
+static zend_never_inline void zend_string_table_grow(HashTable *interned_strings)
+{
+	if (EXPECTED(interned_strings->nTableSize < HT_MAX_SIZE)) {	/* Let's double the table size */
+		void *new_data;
+		void *old_data = HT_GET_DATA_ADDR(interned_strings);
+		Bucket *old_buckets = interned_strings->arData;
+
+		interned_strings->nTableSize += interned_strings->nTableSize;
+		interned_strings->nTableMask = -interned_strings->nTableSize;
+		new_data = pemalloc(HT_SIZE(interned_strings), interned_strings->u.flags & HASH_FLAG_PERSISTENT);
+
+		if (EXPECTED(new_data)) {
+			HT_SET_DATA_ADDR(interned_strings, new_data);
+			memcpy(interned_strings->arData, old_buckets, sizeof(Bucket) * interned_strings->nNumUsed);
+			pefree(old_data, interned_strings->u.flags & HASH_FLAG_PERSISTENT);
+			zend_hash_rehash(interned_strings);
+		} else {
+			interned_strings->nTableSize = interned_strings->nTableSize >> 1;
+			interned_strings->nTableMask = -interned_strings->nTableSize;
 		}
 	}
+}
+
+/* This function might be not thread safe at least because it would update the
+   hash val in the passed string. Be sure it is called in the appropriate context. */
+static zend_always_inline zend_string *zend_add_interned_string(zend_string *str, HashTable *interned_strings, uint32_t flags)
+{
+	zend_ulong h;
+	uint32_t nIndex;
+	uint32_t idx;
+	Bucket *p;
+
+	h = zend_string_hash_val(str);
+
+	GC_REFCOUNT(str) = 1;
+	GC_FLAGS(str) |= IS_STR_INTERNED | flags;
+
+	if (UNEXPECTED(interned_strings->nNumUsed >= interned_strings->nTableSize)) {
+		zend_string_table_grow(interned_strings);
+	}
 
-	idx = CG(interned_strings).nNumUsed++;
-	CG(interned_strings).nNumOfElements++;
-	p = CG(interned_strings).arData + idx;
+	idx = interned_strings->nNumUsed++;
+	interned_strings->nNumOfElements++;
+	p = interned_strings->arData + idx;
 	p->h = h;
 	p->key = str;
 	Z_STR(p->val) = str;
 	Z_TYPE_INFO(p->val) = IS_INTERNED_STRING_EX;
-	nIndex = h | CG(interned_strings).nTableMask;
-	Z_NEXT(p->val) = HT_HASH(&CG(interned_strings), nIndex);
-	HT_HASH(&CG(interned_strings), nIndex) = HT_IDX_TO_HASH(idx);
+	nIndex = h | interned_strings->nTableMask;
+	Z_NEXT(p->val) = HT_HASH(interned_strings, nIndex);
+	HT_HASH(interned_strings, nIndex) = HT_IDX_TO_HASH(idx);
```

Comment:
This could use `zend_hash_append()`. Or going one step further, this could just use `zend_hash_add_new()`, in which case we can discard the duplicated table growing logic as well.